x-pack/elastic-agent/pkg/composable/providers/docker/docker.go

```diff
@@ -0,0 +1,157 @@
+		return nil
+	}
+	startListener := watcher.ListenStart()
+	stopListener := watcher.ListenStop()
+	stoppers := map[string]*time.Timer{}
+	stopTrigger := make(chan *dockerContainerData)
+
+	if err := watcher.Start(); err != nil {
+		// info only; return nil (do nothing)
+		c.logger.Infof("Docker provider skipped, unable to connect: %s", err)
+		return nil
+	}
+
+	go func() {
+		for {
+			select {
+			case <-comm.Done():
+				startListener.Stop()
+				stopListener.Stop()
+
+				// Stop all timers before closing the channel
+				for _, stopper := range stoppers {
+					stopper.Stop()
+				}
+				close(stopTrigger)
+				return
+			case event := <-startListener.Events():
+				data, err := generateData(event)
+				if err != nil {
+					c.logger.Errorf("%s", err)
+					continue
+				}
+				if stopper, ok := stoppers[data.container.ID]; ok {
+					c.logger.Debugf("container %s is restarting, aborting pending stop", data.container.ID)
+					stopper.Stop()
+					delete(stoppers, data.container.ID)
+					return
+				}
+				comm.AddOrUpdate(data.container.ID, data.mapping, data.processors)
+			case event := <-stopListener.Events():
+				data, err := generateData(event)
+				if err != nil {
+					c.logger.Errorf("%s", err)
+					continue
+				}
+				stopper := time.AfterFunc(c.config.CleanupTimeout, func() {
+					stopTrigger <- data
+				})
+				stoppers[data.container.ID] = stopper
+			case data := <-stopTrigger:
+				if _, ok := stoppers[data.container.ID]; ok {
+					delete(stoppers, data.container.ID)
+				}
+				comm.Remove(data.container.ID)
+			}
+		}
+	}()
+
+	return nil
+}
+
+// DynamicProviderBuilder builds the dynamic provider.
+func DynamicProviderBuilder(c *config.Config) (composable.DynamicProvider, error) {
+	logger, err := logger.New("composable.providers.docker")
+	if err != nil {
+		return nil, err
+	}
+	var cfg Config
+	if c == nil {
+		c = config.New()
+	}
+	err = c.Unpack(&cfg)
+	if err != nil {
+		return nil, errors.New(err, "failed to unpack configuration")
+	}
+	return &dynamicProvider{logger, &cfg}, nil
+}
+
+func generateData(event bus.Event) (*dockerContainerData, error) {
+	container, ok := event["container"].(*docker.Container)
+	if !ok {
+		return nil, fmt.Errorf("unable to get container from watcher event")
+	}
+
+	labelMap := common.MapStr{}
+	processorLabelMap := common.MapStr{}
+	for k, v := range container.Labels {
+		safemapstr.Put(labelMap, k, v)
+		processorLabelMap.Put(common.DeDot(k), v)
+	}
+
+	data := &dockerContainerData{
+		container: container,
+		mapping: map[string]interface{}{
+			"container": map[string]interface{}{
+				"id":     container.ID,
+				"name":   container.Name,
+				"image":  container.Image,
+				"labels": labelMap,
+			},
+		},
+		processors: []map[string]interface{}{
+			{
+				"add_fields": map[string]interface{}{
+					"fields": map[string]interface{}{
+						"id":     container.ID,
+						"name":   container.Name,
+						"image":  container.Image,
+						"labels": processorLabelMap,
```

Comment:
We should revisit later on what the default info is that we ship, if we potentially should reduce it.

Reply:
Works for me.

Reply:
@blakerouse Could you follow up with an issue? I'm pretty sure we will forget otherwise :-D

Reply:
Filed #21135